### ISSUE TITLE
crash in adios2_open

### ISSUE THREAD

**Issue by germasch** [ornladios/ADIOS2#2114]
I've been working on debugging an issue with a new checkpointing implementation in the GENE code that uses adios2. I made a (somewhat) minimal reproducer below, that first creates the file, and then opens it back for reading. It crashes in read() when opening the file (`adios2_open`). with an nondescript segfault:

```
root@30678029a2d4:/src/gene-dev/build-spack-ubuntu# src/adios2_test

Program received signal SIGSEGV: Segmentation fault - invalid memory reference.

Backtrace for this error:
#0  0x7fea5669a2ed in ???
#1  0x7fea56699503 in ???
#2  0x7fea562ccf1f in ???
#3  0x7fea5512defa in ???
#4  0x7fea55cbf3e8 in ???
#5  0x7fea55cb10ff in ???
#6  0x7fea55cb12b4 in ???
#7  0x7fea55b41e76 in ???
#8  0x7fea55b42cd0 in ???
#9  0x7fea55b48df2 in ???
#10  0x7fea55a80366 in ???
#11  0x7fea55ea900a in ???
#12  0x7fea56d219f0 in ???
#13  0x7fea56d2385d in ???
#14  0x55e4b9208399 in __cp_MOD_read
	at /src/gene-dev/src/adios2_test.F90:56
#15  0x55e4b9207eb1 in test
	at /src/gene-dev/src/adios2_test.F90:77
#16  0x55e4b9207eb1 in main
	at /src/gene-dev/src/adios2_test.F90:67
Segmentation fault
```

This is using adios@2.5.0 compiled with Spack, in an Ubuntu 18.04 container.

```
module cp
  integer :: arr_of_sizes(6) = [10, 10, 10, 10, 10, 10]
  integer :: arr_of_starts(6) = [0, 0, 0, 0, 0, 0]
  integer :: arr_of_subsizes(6) = [5, 5, 5, 5, 5, 5]

contains
  subroutine write()
    use adios2
    use mpi
    implicit none
    
    integer :: ierr
    type(adios2_adios) :: adios_o
    type(adios2_io) :: io
    type(adios2_engine) :: engine
    type(adios2_variable) :: var
    real(kind=8) :: g(5, 5, 5, 5, 5, 5)
  
    call adios2_init(adios_o, MPI_COMM_WORLD, .true., ierr)
    
    call adios2_declare_io(io, adios_o, "checkpoint_write", ierr)
    call adios2_define_variable(var, io, "g_", adios2_type_dp, 6, &
         int(arr_of_sizes,kind=8),int( arr_of_starts,kind=8),&
         int(arr_of_subsizes,kind=8), .true., ierr)
    
    g = 99
    call adios2_open(engine, io, "checkpoint.bp", adios2_mode_write, &
         MPI_COMM_WORLD, ierr)
    call adios2_put(engine, "g_", g, ierr)
    call adios2_close(engine, ierr)
    
    call adios2_finalize(adios_o, ierr)

  end subroutine write

  subroutine read()
    use adios2
    use mpi
    implicit none
    
    integer :: ierr
    type(adios2_adios) :: adios_o
    type(adios2_io) :: io
    type(adios2_engine) :: engine
    type(adios2_variable) :: var
    
    call adios2_init(adios_o, MPI_COMM_WORLD, .true., ierr)
    
    call adios2_declare_io(io, adios_o, "checkpoint_read", ierr)
    call adios2_define_variable(var, io, "g_", adios2_type_dp, 6, &
         & int(arr_of_sizes,kind=8),int( arr_of_starts,kind=8),&
         & int(arr_of_subsizes,kind=8), .true., ierr)
    
    call adios2_open(engine, io, "checkpoint.bp", adios2_mode_read, &
       MPI_COMM_WORLD, ierr)
    ! CRASHES HERE
    
    call adios2_finalize(adios_o, ierr)
    
  end subroutine read
  
end module cp


program test
  use adios2
  use mpi
  use cp
  implicit none

  integer :: ierr

  call mpi_init(ierr)

  call write()
  call read()

  call mpi_finalize(ierr)

end program test
```

**Comment by pnorbert** [comment]
Kai,
The reason for the abort is that you define the variable "g_" for the io group, then you open a file which has this variable. A meaningful error message would have been better instead of the segfault of course.
On the read side, you don't need (and shouldn't, apparently) to define the variables. Open() will create all variables in IO for you. After that you can do adios2_inquire_variable() to get to one of them if needed. 

```  subroutine read()
    use adios2
    use mpi
    implicit none

    integer :: ierr
    type(adios2_adios) :: adios_o
    type(adios2_io) :: io
    type(adios2_engine) :: engine
    type(adios2_variable) :: var
    real(kind=8) :: g(5, 5, 5, 5, 5, 5)
    integer*8, dimension(6) :: start, count

    start = (/ 0, 0, 0, 0, 0, 0 /)
    count = (/ 5, 5, 5, 5, 5, 5 /)

    call adios2_init(adios_o, MPI_COMM_WORLD, .true., ierr)

    call adios2_declare_io(io, adios_o, "checkpoint_read", ierr)
    call adios2_open(engine, io, "checkpoint.bp", adios2_mode_read, &
       MPI_COMM_WORLD, ierr)
    call adios2_inquire_variable(var, io, "g_", ierr)
    call adios2_set_selection(var, 6, start, count, ierr)
    call adios2_get(engine, var, g, ierr)
    call adios2_close(engine, ierr)
    call adios2_finalize(adios_o, ierr)

  end subroutine read
```

**Comment by williamfgc** [comment]
@pnorbert what's strange is that is crashing in `adios2_open` as it should return `ierr` exposing the exception to error conversion from C++. The exception message is piped to `cerr`. @germasch do you think the crash might be happening at `adios2_put`?

**Comment by pnorbert** [comment]
I found it strange too, but it crashes in adios2_open when parsing the metadata and trying to define the g_ variable. I did not dive deeper as I realized the fix is at the user level.

**Comment by germasch** [comment]
Thanks, that I fixed, and in fact I should have remembered, since I had made the same mistake quite a while ago, though with the C++ bindings. (This time I didn't write the actual code, I'm just working on integrating it upstream).

Back then, I got an explicit error from adios2, though, rather than a crash (something like "variable already defined" , IIRC).

**Comment by pnorbert** [comment]
@williamfgc The original example crashes with invalid memory access segfault in parsing the variable metadata. Not an exception thrown by the core. 

BP4Deserializer.tcc:849
```
                    if (subsetCharacteristics.Shape !=
                        variable->m_AvailableShapes.rbegin()->second)
                    {
                        variable->m_AvailableShapes[step] =
                            subsetCharacteristics.Shape;
                    }
```

The variable is defined **before** Open and `m_AvailableShapes` is an empty map.

**Comment by williamfgc** [comment]
@pnorbert thanks for checking. Yes, checking the `rbegin` (or any iterator) against `end` iterators before dereferencing `second` should do the trick.

``` 
auto itrBegin = variable->m_AvailableShapes.rbegin() ;
if( itrBegin != variable->m_AvailableShapes.end() ) 
{
   ...
```

**Comment by germasch** [comment]
Yeah, so I had actually tried the same fix as @williamfgc suggested, and it fixed the crash. But it unfortunately also still lead to a messed up internal state where now I wouldn't get an error, but then actually reading the variable would fail, IIRC.

I think the preferred behavior would be to abort with an error message telling the user what they did wrong. The second best is the current behavior, ie., a crash, so at least the user knows something is wrong, though it's not clear what.
I think being silent about being used incorrectly but then not working right subsequently is the least desirable way of treating it.

So my suggestion would be something of this sort:
```
if (variable->m_AvailableShapes.empty()) {
   // throw error "BUG: m_AvailableShapes is unexpectedly empty. Did you call defineVariable() before opening a file for reading?"
}
```

It might be better if this kind of user mistake could be detected higher up, but this at least would deal with the particular issue I encountered.
